snps: nsim: arc_v: add RHX100 support

AFOliveira opened the pull request:
Add RHX100 RISC-V board configuration for ARC-V cores with the following changes:

- soc: snps: nsim: arc_v: rhx: add RHX SoC configuration
  Add RISC-V RHX SoC series configuration for ARC-V RHX cores. Enables
  RV32IMAC ISA with bitmanip extensions (ZBA, ZBB, ZBC, ZBS). Configures PMP
  with 16 slots and 8-byte granularity. Sets RISCV_SOC_INTERRUPT_INIT enabled
  for interrupt initialization. Configures 32 IRQs. Adds MetaWare CCAC
  toolchain support with RHX-specific compiler flags.

- boards: snps: nsim: arc_v: add RHX100 board configuration
  Add board configuration for RHX100 RISC-V processor. Provides device tree
  files for RHX100 and RHX1xx series, YAML configuration for nSIM simulator
  integration, and board-specific default configuration.

- doc: boards: nsim_arc_v: add RHX100 to board documentation
  Add RHX100 board target to the nsim_arc_v platform documentation. Update
  supported boards list and toolchain checking examples to include both
  RMX100 and RHX100 references.

To get conformity with RMX, this PR introduces a small change on RMX that
does the following:
- Add explicit PMP_GRANULARITY default value of 8 to RMX SoC configuration
  to match the RHX configuration. This is purely for verbosity – no
  functional change, as RISCV_PMP was already enabled and PMP_GRANULARITY
  already defaulted to this value.

AFOliveira committed:
soc: snps: nsim: arc_v: rhx: add RHX SoC configuration

Add RISC-V RHX SoC series configuration for ARC-V RHX cores.
Enables RV32IMAC ISA with bitmanip extensions (ZBA, ZBB, ZBC, ZBS).
Configures PMP with 16 slots and 8-byte granularity.
Sets RISCV_SOC_INTERRUPT_INIT enabled for interrupt initialization.
Configures 32 IRQs.
Adds MetaWare CCAC toolchain support with RHX-specific compiler flags.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
soc: snps: nsim: arc_v: rmx: add PMP_GRANULARITY configuration

Add explicit PMP_GRANULARITY default value of 8 to RMX SoC configuration.
This makes the PMP configuration more explicit and easier to understand.

No functional change as RISCV_PMP was already enabled and PMP_GRANULARITY
already defaulted to this value.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
doc: boards: nsim_arc_v: add RHX100 to board documentation

Add RHX100 board target to the nsim_arc_v platform documentation.
Update supported boards list and toolchain checking examples to
include both RMX100 and RHX100 references.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira committed:
boards: snps: nsim: arc_v: add RHX100 board configuration

Add board configuration for RHX100 RISC-V processor.
Provides device tree files for RHX100 and RHX1xx series,
YAML configuration for nSIM simulator integration,
and board-specific default configuration.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

nordicjm commented on soc/snps/nsim/arc_v/rhx/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if(COMPILER STREQUAL arcmwdt)
+# CCAC:
+  zephyr_compile_options(-av5rhx -Zicsr -Zifencei -Zihintpause -Zba -Zbb
+                         -Zbc -Zbs -Zca -Zcb -Zcmp -Zcmt -Za -Zm  -Zc
+                         -Zicond -Zicbom -Zicbop)
```
checked with Torsten, these should not be here and should be defined by toolchain and just referencing those properties

tagunil replied:
Hi @nordicjm, in theory you're right, but we don't support mapping RISC-V extensions to compiler flags for our MWDT toolchain just yet. We do have such a task in our backlog though.

ruuddw replied:
This is consistent with the way it is done for Metaware, also for ARC Classic targets. A better solution can be done later I think (e.g. using tcf files, or using the ISA feature kconfigs like for RISC-V gcc toolchain).

tejlmand replied:
but it's not consistent with how we do toolchain support.

One reason for the whole toolchain / compiler / linker infrastructure and property scheme was to be able for developers to just write code and not have to worry if this code is to be compiled with gcc, clang, arcmwdt, iar, or any other compiler.

Just apply the compiler property and the it's for the integration of current compiler to decide how to map requested property to flag (or no flag if not supported or supported differently).

Just because no one has noticed this issue doesn't make it correct.
It should be cleaned up before other starts copying this principle elsewhere in Zephyr.

ARC Metaware toolchain developers ought to know this, as they are actually benefiting a lot from this design because it makes it much easier for them to support all samples in Zephyr and not having to go through Zephyr's CMake code file by file because they would need to use a flag different from gcc.

I'm not familiar with the different flags here, like `-Zbc -Zbs -Zca -Zcb -Zcmp -Zcmt -Za -Zm -Zc` etc.
but I assume those has to do with the core, similar to what is achieved with gcc/clang when using `-mcpu`, correct ?

AFOliveira replied:
Should be fixed.

tagunil commented on soc/snps/nsim/arc_v/rhx/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+	select SOC_FAMILY_NSIM_ARC_V
+
+config SOC_SERIES
+	default "rhx" if SOC_SERIES_NSIM_ARC_V_RHX
```
Is this a hard requirement? We've seen a different approach, for example, in https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/andestech/ae350/Kconfig.soc#L23, or even in https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nordic/Kconfig.soc#L11

nordicjm replied:
It needs to be yes, those were mistakes left in with the original port to hwmv2. There is a future plan to make SoC selection fully automatic in CMake, and if the names do not match, boards/socs will then fail

AFOliveira committed:
dts: cpu: add device tree bindings for Synopsys ARC-V RMX RISC-V CPU

Add device tree binding file for the Synopsys ARC-V RMX RISC-V
CPU core. This binding enables proper device tree property parsing by
the Enhanced Device Tree (EDT) system, allowing Kconfig device tree
macros to access CPU properties like clock-frequency.

The binding includes the standard RISC-V CPU properties by extending
riscv,cpus.yaml, which provides access to properties defined in cpu.yaml
such as clock-frequency.

Also removes hardcoded SYS_CLOCK_HW_CYCLES_PER_SEC from board level
and adds DT-derived value to RMX SoC level.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

AFOliveira commented on boards/snps/nsim/arc_v/Kconfig.defconfig:
```diff
@@ -10,3 +10,13 @@ config SYS_CLOCK_HW_CYCLES_PER_SEC
+config SYS_CLOCK_TICKS_PER_SEC
+	default 10000
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 5000000
```
Done, and moved to SoC as per other SoCs descriptions.

AFOliveira committed:
soc: snps: nsim: arc_v: rmx: revert to shorter SOC configuration names

Uses shorter names for the RMX SOC configuration.

Changes:
- SOC_SERIES_NSIM_ARC_V_RMX -> SOC_SERIES_RMX
- SOC_NSIM_ARC_V_RMX100 -> SOC_RMX100

Also updates board Kconfig to use the shorter names.

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

tejlmand reviewed (changes requested):
This seems to continue on a wrong direction wrt. applying compiler specific flags based on actual compiler, and thereby bypassing the compiler infrastructure design.

Please correct, or explain why deviation from the design principle of toolchain infrastructure should be accepted.

AFOliveira committed:
cmake: arcmwdt: riscv: derive ccac flags from Kconfig

Add arcmwdt RISC-V core/ISA mapping in target_riscv.cmake to
automatically derive compiler flags from Kconfig settings.

- Select series core flag from SoC Kconfig: -av5rmx/-av5rhx
- Map RISC-V Kconfig to ccac -Z flags (M/A/C/Zicsr/Zifencei/Zicntr)
- Handle compressed logic correctly: -Zc or individual Zc* sub-extensions
- Add F/D and compressed FPU (-Zf/-Zd, -Zcf/-Zcd)
- Add -Zmmul when M is not selected and ZMMUL is enabled
- Use OPTIONAL include to allow for arch-specific target files

Keep early compiler checks stable by not pushing -Z flags into
CMAKE_REQUIRED_FLAGS.

Co-authored-by: Torsten Tejlmand Rasmussen <torsten.rasmussen@nordicsemi.no>
Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>
Signed-off-by: Afonso Oliveira <afonso.oliveira707@gmail.com>
Signed-off-by: Torsten Tejlmand Rasmussen <torsten.rasmussen@nordicsemi.no>

AFOliveira committed:
soc: snps: nsim: arc_v: rmx: drop hardcoded MWDT flags

- Remove SoC-level ccac core/ISA flags (-av5rmx, -Z*)
- Depend on arcmwdt toolchain to derive flags from Kconfig
- Keep SOC linker script selection unchanged

Signed-off-by: Afonso Oliveira <afonsoo@synopsys.com>

ruuddw commented on boards/snps/nsim/arc_v/Kconfig.defconfig:
```diff
@@ -6,7 +6,11 @@ if BOARD_NSIM_ARC_V_RMX100
+if BOARD_NSIM_ARC_V_RHX100
+
+config SYS_CLOCK_TICKS_PER_SEC
+	default 10000
```
10000 is a very high number for this. 100 or 1000 is the normal value. With tickless kernel the higher value is probably ok, but without that's a lot of time ISR processing...

ruuddw reviewed (changes requested):
Please reduce the ticks per sec (or explain why it is set so high).

tejlmand commented on cmake/compiler/arcmwdt/target_riscv.cmake:
The general pattern is to set define the flags in `compiler_flags.cmake` for the compiler and then refer to those where needed, but given the nature of the arc mwdt flags and their corresponding Kconfig settings, then the approach used in this case is accepted.

tejlmand commented on cmake/compiler/arcmwdt/target.cmake:
```diff
@@ -63,3 +63,8 @@ endif()
 endfunction()
+
+# Include architecture-specific settings
+if("${ARCH}" STREQUAL "riscv")
+  include(${CMAKE_CURRENT_LIST_DIR}/target_riscv.cmake)
+endif()
```
Hint: You can also do like this and thereby allow to just ad new `target_<arch>.cmake` files when needed without having to add extra `if() ... elseif() ... endif()` in future.
```suggestion
include(${CMAKE_CURRENT_LIST_DIR}/target_${ARCH}.cmake OPTIONAL)
```

tejlmand commented on cmake/compiler/arcmwdt/target_riscv.cmake:
```diff
@@ -0,0 +1,114 @@
+endif()
+
+if(NOT ARCMWDT_RISCV_CORE)
+  message(FATAL_ERROR "MWDT: Unable to determine ARC-V series from SoC Kconfig (expect CONFIG_SOC_SERIES_RMX or CONFIG_SOC_SERIES_RHX, etc.).")
```
Please line wrap.

Ref: https://docs.zephyrproject.org/latest/contribute/style/cmake.html#general-formatting

AFOliveira replied:
Done.

tejlmand reviewed (dismissed):
Much better :+1: 

A single too long lines which should be fixed before merging.
But as not blocking, PRR is approved.

nordicjm commented:
Seems CI needs fixing

AFOliveira commented:
@nordicjm I think this was not created by my PR and plan to rebase soon to see if it goes away. If it's related to my PR I'll fix it ASAP.

ruuddw reviewed (approved):
Just a minor thing I noticed now, some new files have "copyright 2024" - guess you didn't create them last year...
